lib/ansible/galaxy/role.py

```diff
@@ -334,7 +334,7 @@ def install(self):
                                 n_parts = n_member_name.replace(n_archive_parent_dir, "", 1).split(os.sep)
                                 n_final_parts = []
                                 for n_part in n_parts:
-                                    if n_part != '..' and '~' not in n_part and '$' not in n_part:
+                                    if n_part != '..' and n_part != '~' and '$' not in n_part:
```

Comment:
What about `~username/dir` that resolves into `/home/username/dir`? It looks like you need
```suggestion
                                    if n_part != '..' and not n_part.startswith('~') and '$' not in n_part:
```

Also, I'm not sure if you need to check this in all the parts or only the first one.